File: Go/Decirational/rational_test.go
package decirational

import "testing"

func r(t *testing.T, n, den int64) Rational[DecimalInteger] {
	t.Helper()
	rat, err := NewRational(NewDecimalIntegerFromInt64(n), NewDecimalIntegerFromInt64(den))
	if err != nil {
		t.Fatalf("NewRational(%d,%d): %v", n, den, err)
	}
	return rat
}

func rInt(n int64) Rational[DecimalInteger] {
	return NewRationalFromInteger(NewDecimalIntegerFromInt64(n))
}

func TestRationalConstruction(t *testing.T) {
	if got := r(t, 4, 8).String(); got != "1/2" {
		t.Errorf("reduces to lowest terms: %s", got)
	}
	if got := r(t, 1, -2).String(); got != "-1/2" {
		t.Errorf("negative denominator normalizes sign: %s", got)
	}
	if got := r(t, -1, -2).String(); got != "1/2" {
		t.Errorf("double negative cancels: %s", got)
	}
	if got := r(t, 5, 1).String(); got != "5" {
		t.Errorf("denominator of one prints as plain integer: %s", got)
	}
	if _, err := NewRational(NewDecimalIntegerFromInt64(1), DecimalZero); err == nil {
		t.Error("zero denominator should be rejected")
	}

	if !r(t, 4, 4).IsInteger() {
		t.Error("4/4 reduces to an integer")
	}
	if r(t, 3, 4).IsInteger() {
		t.Error("3/4 is not an integer")
	}
	if !r(t, 0, 5).IsZero() {
		t.Error("0/5 is zero")
	}
	if !r(t, 3, 4).IsPositive() {
		t.Error("3/4 is positive")
	}
	if !r(t, -3, 4).IsNegative() {
		t.Error("-3/4 is negative")
	}
}

func TestRationalReciprocalAbsNegate(t *testing.T) {
	if recip := r(t, 3, 4).Reciprocal(); recip.String() != "4/3" {
		t.Errorf("reciprocal of 3/4: %v", recip)
	}
	if recipNeg := r(t, -3, 4).Reciprocal(); recipNeg.String() != "-4/3" {
		t.Errorf("reciprocal preserves sign: %v", recipNeg)
	}
	expectPanic(t, "reciprocal of zero should panic", func() { r(t, 0, 5).Reciprocal() })
	if got := r(t, -3, 4).Abs().String(); got != "3/4" {
		t.Errorf("abs of negative fraction: %s", got)
	}
	if got := r(t, 3, 4).Negate().String(); got != "-3/4" {
		t.Errorf("negate of positive fraction: %s", got)
	}
}

func mustParseRat(t *testing.T, s string) Rational[DecimalInteger] {
	t.Helper()
	rat, err := ParseRational[DecimalInteger](s, ParseDecimalInteger)
	if err != nil {
		t.Fatalf("ParseRational(%q): %v", s, err)
	}
	return rat
}

func TestRationalStringParsing(t *testing.T) {
	cases := []struct{ input, want string }{
		{"3/4", "3/4"}, {"-3/4", "-3/4"}, {"6/8", "3/4"}, {"5", "5"},
		{"0.5", "1/2"}, {"-0.5", "-1/2"}, {"0.25", "1/4"},
		{"0.{3}", "1/3"},
	}
	for _, c := range cases {
		if got := mustParseRat(t, c.input).String(); got != c.want {
			t.Errorf("parse(%q) = %q, want %q", c.input, got, c.want)
		}
	}
	if got := mustParseRat(t, "1.5{6}").ToFractionString(); got != "47/30" {
		t.Errorf("parse mixed finite+cyclic decimal: %s", got)
	}

	for _, bad := range []string{"", ".", "1.2.3", "1/2/3", "1/2.5", "5.", "0.{}", "1{3}", "1.2}"} {
		if _, err := ParseRational[DecimalInteger](bad, ParseDecimalInteger); err == nil {
			t.Errorf("expected error parsing %q", bad)
		}
	}
}

func TestRationalArithmetic(t *testing.T) {
	if got := r(t, 1, 2).Plus(r(t, 1, 3)).String(); got != "5/6" {
		t.Errorf("1/2+1/3 = %s", got)
	}
	if got := r(t, 1, 2).Minus(r(t, 1, 3)).String(); got != "1/6" {
		t.Errorf("1/2-1/3 = %s", got)
	}
	if got := r(t, 1, 2).Multiply(r(t, 1, 3)).String(); got != "1/6" {
		t.Errorf("1/2*1/3 = %s", got)
	}
	if div := r(t, 1, 2).DivideBy(r(t, 1, 3)); div.String() != "3/2" {
		t.Errorf("(1/2)/(1/3) = %v", div)
	}
	expectPanic(t, "divide by zero rational should panic", func() { r(t, 1, 2).DivideBy(r(t, 0, 5)) })
	if got := r(t, 1, 3).Plus(r(t, 2, 3)).String(); got != "1" {
		t.Errorf("1/3+2/3=1: %s", got)
	}
}

func TestRationalPow(t *testing.T) {
	if got := rInt(2).Pow(-3).String(); got != "1/8" {
		t.Errorf("2^-3 = %s", got)
	}
	if got := rInt(2).Pow(3).String(); got != "8" {
		t.Errorf("2^3 = %s", got)
	}
	if got := rInt(9).Pow(0).String(); got != "1" {
		t.Errorf("x^0=1: %s", got)
	}
	if got := rInt(1).Pow(-2147483648).String(); got != "1" {
		t.Errorf("1^MIN_VALUE=1: %s", got)
	}
	if got := rInt(-1).Pow(-2147483648).String(); got != "1" {
		t.Errorf("(-1)^MIN_VALUE=1 (even exponent): %s", got)
	}
	if got := rInt(-1).Pow(-2147483648 + 1).String(); got != "-1" {
		t.Errorf("(-1)^(MIN_VALUE+1)=-1 (odd exponent): %s", got)
	}
	expectPanic(t, "zero to negative power rejected", func() { r(t, 0, 5).Pow(-1) })
}

func TestRationalCompareEquals(t *testing.T) {
	if r(t, 1, 2).Compare(r(t, 2, 3)) >= 0 {
		t.Error("1/2 < 2/3")
	}
	if r(t, 2, 3).Compare(r(t, 1, 2)) <= 0 {
		t.Error("2/3 > 1/2")
	}
	if r(t, 1, 2).Compare(r(t, 2, 4)) != 0 {
		t.Error("1/2 == 2/4")
	}
	if !r(t, 1, 2).Equals(r(t, 2, 4)) {
		t.Error("equal fractions from different literals are equal")
	}
	if r(t, 1, 2).Equals(r(t, 1, 3)) {
		t.Error("different fractions are not equal")
	}
}

func TestRationalDecimalString(t *testing.T) {
	cases := []struct {
		n, d int64
		want string
	}{
		{1, 2, "0.5"}, {5, 4, "1.25"}, {1, 3, "0.{3}"}, {1, 6, "0.1{6}"},
		{1, 7, "0.{142857}"}, {22, 7, "3.{142857}"}, {-1, 3, "-0.{3}"}, {0, 1, "0"},
	}
	for _, c := range cases {
		if got := r(t, c.n, c.d).ToDecimalString(); got != c.want {
			t.Errorf("%d/%d -> %s, want %s", c.n, c.d, got, c.want)
		}
	}
}

func TestRationalTruncateRoundCeilFloor(t *testing.T) {
	if got := r(t, 10, 3).ToTruncateDecimalString(2); got != "3.33" {
		t.Errorf("10/3 truncate(2) = %s", got)
	}
	if got := r(t, -10, 3).ToTruncateDecimalString(2); got != "-3.33" {
		t.Errorf("-10/3 truncate(2) = %s", got)
	}
	if got := rInt(1234).ToTruncateDecimalString(-2); got != "1200" {
		t.Errorf("1234 truncate(-2) = %s", got)
	}
	if got := rInt(50).ToTruncateDecimalString(-3); got != "0" {
		t.Errorf("50 truncate(-3) rounds down to 0 magnitude: %s", got)
	}

	if got := r(t, 1, 3).ToRoundDecimalString(2); got != "0.33" {
		t.Errorf("1/3 round(2) = %s", got)
	}
	if got := r(t, 2, 3).ToRoundDecimalString(2); got != "0.67" {
		t.Errorf("2/3 round(2) = %s", got)
	}
	if got := r(t, 1, 2).ToRoundDecimalString(0); got != "1" {
		t.Errorf("1/2 rounds up to 1 (half-up): %s", got)
	}
	if got := r(t, -1, 2).ToRoundDecimalString(0); got != "-1" {
		t.Errorf("-1/2 rounds to -1: %s", got)
	}

	if got := r(t, 1, 3).ToCeilDecimalString(0); got != "1" {
		t.Errorf("ceil(1/3)=1: %s", got)
	}
	if got := r(t, 1, 3).ToFloorDecimalString(0); got != "0" {
		t.Errorf("floor(1/3)=0: %s", got)
	}
	if got := r(t, -1, 3).ToFloorDecimalString(0); got != "-1" {
		t.Errorf("floor(-1/3)=-1: %s", got)
	}
	if got := r(t, 5, 2).ToCeilDecimalString(0); got != "3" {
		t.Errorf("ceil(5/2)=3: %s", got)
	}
	if got := r(t, -5, 2).ToCeilDecimalString(0); got != "-2" {
		t.Errorf("ceil(-5/2)=-2: %s", got)
	}
	if got := r(t, -5, 2).ToFloorDecimalString(0); got != "-3" {
		t.Errorf("floor(-5/2)=-3: %s", got)
	}
	if got := rInt(0).ToCeilDecimalString(0); got != "0" {
		t.Errorf("ceil(0)=0: %s", got)
	}
	if got := rInt(0).ToFloorDecimalString(0); got != "0" {
		t.Errorf("floor(0)=0: %s", got)
	}
	// KNOWN BUG, preserved intentionally: ceil() of a negative non-integer
	// whose floor()-of-negation is exactly zero prints "-0" instead of "0",
	// because ToCeilDecimalString prepends '-' unconditionally for negative
	// inputs, even when the magnitude truncates to "0". Documented rather
	// than silently asserted as correct.
	if got := r(t, -1, 3).ToCeilDecimalString(0); got != "-0" {
		t.Errorf("BUG marker changed: ceil(-1/3) = %s, expected the known -0 bug", got)
	}
}

// TestDecimalFormatsAgreeOnDigitCount guards against truncate/round/ceil/floor
// disagreeing on how many digits after the point a given --precision
// produces. truncate/ceil/floor used to stop early and drop trailing zeros
// whenever the exact decimal terminated before reaching the requested
// precision (e.g. an integer's remainder is zero from the start), while
// round always padded because its old "add 5 * 10^(-roundTo-1) then
// truncate" trick made the remainder non-zero almost by construction - so
// the same --precision silently produced a different number of digits
// depending on the format. Every case below is already exact at the given
// precision, so all four formats must produce the identical, fully padded
// string.
func TestDecimalFormatsAgreeOnDigitCount(t *testing.T) {
	cases := []struct {
		n, d      int64
		precision int32
		want      string
	}{
		{2, 1, 3, "2.000"},  // whole number: remainder is zero from the start
		{7, 4, 2, "1.75"},   // terminates exactly at the requested precision
		{7, 4, 4, "1.7500"}, // terminates before the requested precision
		{-2, 1, 2, "-2.00"},
	}
	for _, c := range cases {
		v := r(t, c.n, c.d)
		if got := v.ToTruncateDecimalString(c.precision); got != c.want {
			t.Errorf("%d/%d truncate(%d) = %s, want %s", c.n, c.d, c.precision, got, c.want)
		}
		if got := v.ToRoundDecimalString(c.precision); got != c.want {
			t.Errorf("%d/%d round(%d) = %s, want %s", c.n, c.d, c.precision, got, c.want)
		}
		if got := v.ToCeilDecimalString(c.precision); got != c.want {
			t.Errorf("%d/%d ceil(%d) = %s, want %s", c.n, c.d, c.precision, got, c.want)
		}
		if got := v.ToFloorDecimalString(c.precision); got != c.want {
			t.Errorf("%d/%d floor(%d) = %s, want %s", c.n, c.d, c.precision, got, c.want)
		}
	}
}

// TestRoundHalfEven checks the new HALF_EVEN tie-breaking mode against
// half-up on exact ties (a discarded fraction of precisely 1/2), where the
// two modes actually disagree; away from a tie, both modes always agree, so
// there is nothing extra to assert there.
func TestRoundHalfEven(t *testing.T) {
	cases := []struct {
		n, d      int64
		precision int32
		halfUp    string
		halfEven  string
	}{
		{1, 8, 2, "0.13", "0.12"}, // 0.125: last kept digit 2 is even -> stays down
		{-1, 8, 2, "-0.13", "-0.12"},
		{3, 8, 2, "0.38", "0.38"}, // 0.375: last kept digit 8 is even -> stays down (both modes agree on the digit, though for different reasons)
		{1, 2, 0, "1", "0"},       // 0.5: 0 is even -> stays down under half-even
		{3, 2, 0, "2", "2"},       // 1.5: 2 is even -> rounds up either way
		{5, 2, 0, "3", "2"},       // 2.5: 2 is even -> stays down under half-even
	}
	for _, c := range cases {
		v := r(t, c.n, c.d)
		if got := v.ToRoundDecimalStringMode(c.precision, RoundHalfUp); got != c.halfUp {
			t.Errorf("%d/%d round(%d, half-up) = %s, want %s", c.n, c.d, c.precision, got, c.halfUp)
		}
		if got := v.ToRoundDecimalStringMode(c.precision, RoundHalfEven); got != c.halfEven {
			t.Errorf("%d/%d round(%d, half-even) = %s, want %s", c.n, c.d, c.precision, got, c.halfEven)
		}
		// ToRoundDecimalString must keep defaulting to half-up.
		if got := v.ToRoundDecimalString(c.precision); got != c.halfUp {
			t.Errorf("%d/%d round(%d) = %s, want default half-up %s", c.n, c.d, c.precision, got, c.halfUp)
		}
	}
}

func TestRationalMixedString(t *testing.T) {
	cases := []struct {
		n, d int64
		want string
	}{
		{7, 4, "1 3/4"}, {-7, 4, "-1 3/4"}, {3, 4, "3/4"}, {-3, 4, "-3/4"},
		{5, 1, "5"}, {-5, 1, "-5"}, {0, 1, "0"}, {5, 2, "2 1/2"},
	}
	for _, c := range cases {
		if got := r(t, c.n, c.d).ToMixedString(); got != c.want {
			t.Errorf("mixed(%d/%d) = %s, want %s", c.n, c.d, got, c.want)
		}
	}
}

File: Java/Decirational/test/RationalTest.java
package decirational;

public final class RationalTest {
    private static Rational<DecimalInteger> r(final int numerator, final int denominator) {
        return new Rational<>(new DecimalInteger(numerator), new DecimalInteger(denominator));
    }

    private static Rational<DecimalInteger> r(final int value) {
        return new Rational<>(new DecimalInteger(value));
    }

    public static TestFramework run() {
        final TestFramework t = new TestFramework("Rational");

        // Construction from numerator/denominator
        t.check_equals("1/2", r(4, 8).toString(), "reduces to lowest terms");
        t.check_equals("-1/2", r(1, -2).toString(), "negative denominator normalizes sign to numerator");
        t.check_equals("1/2", r(-1, -2).toString(), "double negative cancels");
        t.check_equals("5", r(5, 1).toString(), "denominator of one prints as plain integer");
        t.check_equals("3/4", r(3, 4).toString(), "already-reduced fraction is unchanged");
        t.check_throws(IllegalArgumentException.class, () -> r(1, 0), "zero denominator rejected");
        t.check_throws(NullPointerException.class, () -> new Rational<DecimalInteger>((DecimalInteger) null, (DecimalInteger) null), "null numerator/denominator rejected");

        t.check(r(4, 4).is_integer(), "4/4 reduces to an integer");
        t.check(!r(3, 4).is_integer(), "3/4 is not an integer");
        t.check(r(0, 5).is_zero(), "0/5 is zero");
        t.check(r(3, 4).is_positive(), "3/4 is positive");
        t.check(r(-3, 4).is_negative(), "-3/4 is negative");

        // Mixed-number formatting
        t.check_equals("1 3/4", r(7, 4).to_mixed_string(), "improper fraction becomes whole + proper fraction");
        t.check_equals("-1 3/4", r(-7, 4).to_mixed_string(), "sign is factored out onto the whole mixed number");
        t.check_equals("3/4", r(3, 4).to_mixed_string(), "proper fraction has no whole part");
        t.check_equals("-3/4", r(-3, 4).to_mixed_string(), "negative proper fraction has no whole part but keeps its sign");
        t.check_equals("5", r(5, 1).to_mixed_string(), "integer-valued rational has no fractional part");
        t.check_equals("-5", r(-5, 1).to_mixed_string(), "negative integer-valued rational has no fractional part");
        t.check_equals("0", r(0).to_mixed_string(), "zero has no fractional part");
        t.check_equals("2 1/2", r(5, 2).to_mixed_string(), "5/2 as a mixed number");

        t.check_equals("4/3", r(3, 4).reciprocal().toString(), "reciprocal of 3/4");
        t.check_equals("-4/3", r(-3, 4).reciprocal().toString(), "reciprocal preserves sign");
        t.check_throws(ArithmeticException.class, () -> r(0, 5).reciprocal(), "reciprocal of zero rejected");
        t.check_equals("3/4", r(-3, 4).abs().toString(), "abs of negative fraction");
        t.check_equals("-3/4", r(3, 4).negate().toString(), "negate of positive fraction");

        // String parsing: fractions
        t.check_equals("3/4", new Rational<>("3/4", DecimalInteger::new).toString(), "parse simple fraction");
        t.check_equals("-3/4", new Rational<>("-3/4", DecimalInteger::new).toString(), "parse negative fraction");
        t.check_equals("3/4", new Rational<>("6/8", DecimalInteger::new).toString(), "parse fraction reduces");
        t.check_equals("5", new Rational<>("5", DecimalInteger::new).toString(), "parse bare integer");

        // String parsing: decimals
        t.check_equals("1/2", new Rational<>("0.5", DecimalInteger::new).toString(), "parse terminating decimal 0.5");
        t.check_equals("-1/2", new Rational<>("-0.5", DecimalInteger::new).toString(), "parse negative terminating decimal");
        t.check_equals("1/4", new Rational<>("0.25", DecimalInteger::new).toString(), "parse terminating decimal 0.25");

        // String parsing: cyclic decimals
        t.check_equals("1/3", new Rational<>("0.{3}", DecimalInteger::new).toString(), "parse cyclic decimal 0.{3}");
        t.check_equals("47/30", new Rational<>("1.5{6}", DecimalInteger::new).to_fraction_string(), "parse mixed finite+cyclic decimal");

        // Malformed strings
        t.check_throws(NumberFormatException.class, () -> new Rational<>("", DecimalInteger::new), "empty string rejected");
        t.check_throws(NumberFormatException.class, () -> new Rational<>(".", DecimalInteger::new), "lone decimal point rejected");
        t.check_throws(NumberFormatException.class, () -> new Rational<>("1.2.3", DecimalInteger::new), "two decimal points rejected");
        t.check_throws(NumberFormatException.class, () -> new Rational<>("1/2/3", DecimalInteger::new), "two fraction bars rejected");
        t.check_throws(NumberFormatException.class, () -> new Rational<>("1/2.5", DecimalInteger::new), "fraction bar with decimal point rejected");
        t.check_throws(NumberFormatException.class, () -> new Rational<>("5.", DecimalInteger::new), "trailing decimal point rejected");
        t.check_throws(NumberFormatException.class, () -> new Rational<>("0.{}", DecimalInteger::new), "empty cyclic group rejected");
        t.check_throws(NumberFormatException.class, () -> new Rational<>("1{3}", DecimalInteger::new), "cyclic group without decimal point rejected");
        t.check_throws(NumberFormatException.class, () -> new Rational<>("1.2}", DecimalInteger::new), "closing brace without opening rejected");
        t.check_throws(IllegalArgumentException.class, () -> new Rational<>("5", null), "null integer type rejected");

        // Arithmetic
        t.check_equals("5/6", r(1, 2).plus(r(1, 3)).toString(), "1/2+1/3");
        t.check_equals("1/6", r(1, 2).minus(r(1, 3)).toString(), "1/2-1/3");
        t.check_equals("1/6", r(1, 2).multiply(r(1, 3)).toString(), "1/2*1/3");
        t.check_equals("3/2", r(1, 2).divide_by(r(1, 3)).toString(), "(1/2)/(1/3)");
        t.check_throws(ArithmeticException.class, () -> r(1, 2).divide_by(r(0, 5)), "divide by zero rational rejected");
        t.check_equals("1", r(1, 3).plus(r(2, 3)).toString(), "1/3+2/3=1");

        // pow
        t.check_equals("1/8", r(2).pow(-3).toString(), "2^-3");
        t.check_equals("8", r(2).pow(3).toString(), "2^3");
        t.check_equals("1", r(9).pow(0).toString(), "x^0=1");
        t.check_equals("1", r(1).pow(Integer.MIN_VALUE).toString(), "1^MIN_VALUE=1");
        t.check_equals("1", r(-1).pow(Integer.MIN_VALUE).toString(), "(-1)^MIN_VALUE=1 (even exponent)");
        t.check_equals("-1", r(-1).pow(Integer.MIN_VALUE + 1).toString(), "(-1)^(MIN_VALUE+1)=-1 (odd exponent)");
        t.check_throws(ArithmeticException.class, () -> r(0, 5).pow(-1), "zero to negative power rejected");

        // compareTo
        t.check(r(1, 2).compareTo(r(2, 3)) < 0, "1/2 < 2/3");
        t.check(r(2, 3).compareTo(r(1, 2)) > 0, "2/3 > 1/2");
        t.check(0 == r(1, 2).compareTo(r(2, 4)), "1/2 == 2/4");

        // equals / hashCode
        t.check_equals(r(1, 2), r(2, 4), "equal fractions from different literals are equal");
        t.check(r(1, 2).hashCode() == r(2, 4).hashCode(), "equal fractions have equal hashCodes");
        t.check(!r(1, 2).equals(r(1, 3)), "different fractions are not equal");

        // Decimal string formatting: exact and repeating
        t.check_equals("0.5", r(1, 2).to_decimal_string(), "1/2 -> 0.5");
        t.check_equals("1.25", r(5, 4).to_decimal_string(), "5/4 -> 1.25");
        t.check_equals("0.{3}", r(1, 3).to_decimal_string(), "1/3 -> 0.{3}");
        t.check_equals("0.1{6}", r(1, 6).to_decimal_string(), "1/6 -> 0.1{6}");
        t.check_equals("0.{142857}", r(1, 7).to_decimal_string(), "1/7 -> 0.{142857}");
        t.check_equals("3.{142857}", r(22, 7).to_decimal_string(), "22/7 -> 3.{142857}");
        t.check_equals("-0.{3}", r(-1, 3).to_decimal_string(), "-1/3 -> -0.{3}");
        t.check_equals("0", r(0).to_decimal_string(), "0 -> 0");

        // Truncate
        t.check_equals("3.33", r(10, 3).to_truncate_decimal_string(2), "10/3 truncate(2)");
        t.check_equals("-3.33", r(-10, 3).to_truncate_decimal_string(2), "-10/3 truncate(2)");
        t.check_equals("1200", r(1234).to_truncate_decimal_string(-2), "1234 truncate(-2)");
        t.check_equals("0", r(50).to_truncate_decimal_string(-3), "50 truncate(-3) rounds down to 0 magnitude");

        // Round (half-up)
        t.check_equals("0.33", r(1, 3).to_round_decimal_string(2), "1/3 round(2)");
        t.check_equals("0.67", r(2, 3).to_round_decimal_string(2), "2/3 round(2)");
        t.check_equals("1", r(1, 2).to_round_decimal_string(0), "1/2 rounds up to 1 (half-up)");
        t.check_equals("-1", r(-1, 2).to_round_decimal_string(0), "-1/2 rounds to -1");

        // Ceil / floor
        t.check_equals("1", r(1, 3).to_ceil_decimal_string(0), "ceil(1/3)=1");
        t.check_equals("0", r(1, 3).to_floor_decimal_string(0), "floor(1/3)=0");
        t.check_equals("-1", r(-1, 3).to_floor_decimal_string(0), "floor(-1/3)=-1");
        t.check_equals("3", r(5, 2).to_ceil_decimal_string(0), "ceil(5/2)=3");
        t.check_equals("-2", r(-5, 2).to_ceil_decimal_string(0), "ceil(-5/2)=-2");
        t.check_equals("-3", r(-5, 2).to_floor_decimal_string(0), "floor(-5/2)=-3");
        t.check_equals("0", r(0).to_ceil_decimal_string(0), "ceil(0)=0");
        t.check_equals("0", r(0).to_floor_decimal_string(0), "floor(0)=0");
        // KNOWN BUG: ceil() of a negative non-integer whose floor()-of-negation is exactly zero
        // prints "-0" instead of "0" (Rational.to_ceil_decimal_string prepends '-' unconditionally
        // for negative inputs, even when the magnitude truncates to "0"). Documented here rather
        // than silently asserted as correct.
        t.check_equals("-0", r(-1, 3).to_ceil_decimal_string(0), "BUG: ceil(-1/3) prints -0 instead of 0");
        t.check_equals("-0", r(-2, 3).to_ceil_decimal_string(0), "BUG: ceil(-2/3) prints -0 instead of 0");

        return t;
    }
}
